cloudinit/config/cc_growpart.py

```diff
@@ -274,12 +275,69 @@ def devent2dev(devent):
     return dev
 
 
+def is_mapped_device(blockdev) -> bool:
+    """
+    Check if a device is a mapped device.
+
+    blockdev should look something like '/dev/mapper/disk1' if True,
+    otherwise something like /dev/vdb1.
+    """
+    is_mapped = blockdev.startswith('/dev/mapper/')
+    LOG.debug("%s is %s a mapped device", blockdev, "" if is_mapped else "not")
+    return is_mapped
+
+
+def is_encrypted(blockdev) -> bool:
+    """
+    Check if a device is an encrypted device.
+
+    Will work for both mapped and raw devices.
+    """
+    is_encrypted = False
+    if not subp.which('cryptsetup'):
+        LOG.debug('cryptsetup not found. Assuming no encrypted partitions')
+        return False
+    with suppress(subp.ProcessExecutionError):
+        subp.subp(['cryptsetup', 'status', blockdev])
+        is_encrypted = True
+    if not is_encrypted:
+        with suppress(subp.ProcessExecutionError):
+            subp.subp(['cryptsetup', 'isLuks', blockdev])
+            is_encrypted = True
+    LOG.debug(
+        "Determined that %s is %s encrypted",
+        blockdev,
+        "" if is_encrypted else "not"
+    )
+    return is_encrypted
+
+
+def get_underlying_partition(blockdev):
+    command = 'dmsetup deps -o devname {}'.format(blockdev)
+    dep = subp.subp(command.split())[0]
```

Comment:
Generally we should prefer the following for which does not use split or format and uses long format arguments (readability).

```python
dep = subp.subp(["dmsetup", "deps", "--options=devname", blockdev])[0]
```

Also, you really should check the error code... if this returns non-zero, and you try to parse stdout, then you'll likely get an IndexError, and raise a non-helpful exception.

Reply:
Thanks. Was planning some additional error checking in other places too...though...subp will raise error in non-0 exits already. I just don't remember how ugly that is in the logs.